examples/ul2/src/utils/adapt_tokenizer.py

```diff
@@ -0,0 +1,29 @@
+# Copyright 2022 MosaicML Examples authors
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Union
+
+from transformers import PreTrainedTokenizer, PreTrainedTokenizerFast
+
+
+def adapt_tokenizer_for_denoising(tokenizer: Union[PreTrainedTokenizer,
+                                                   PreTrainedTokenizerFast],
+                                  num_sentinel_tokens: int = 100):
+    """Adds sentinel tokens and padding token (if missing).
+
+    Expands the tokenizer vocabulary to include sentinel tokens
+    used in mixture-of-denoiser tasks as well as a padding token.
+
+    All added tokens are added as special tokens. No tokens are
+    added if sentinel tokens and padding token already exist.
+    """
+    # Add sentinel tokens (e.g., <extra_id_0>, <extra_id_1>, and so on). Has no effect if these are already in the vocab.
+    assert 0 <= num_sentinel_tokens <= 1000
+    sentinels_to_add = [f'<extra_id_{i}>' for i in range(num_sentinel_tokens)]
+    tokenizer.add_tokens(sentinels_to_add, special_tokens=True)
+
+    # If the padding token has not been set, add <pad> and use it
+    if tokenizer.pad_token is None:
+        tokenizer.add_tokens('<pad>', special_tokens=True)
+        tokenizer.pad_token = '<pad>'
```

Comment:
is `'<pad>'` universal enough that we can safely hardcode it? (I don't know the answer)

Reply:
This only has an effect if the tokenizer didn't set a pad token for some reason. But there are some style differences here and there. For instance, some prefer `[PAD]`. It's always the tokenizer itself that adds the padding, though, meaning nothing would break if you chose a different convention or even a non-sense string.
So as long as the token is clear to the user (which I argue `<pad>` is), I think it should be fine to make the choice to use `<pad>` if the tokenizer doesn't have a pad token set yet.